kamu-adapter-http/src/ws_smart_transfer_protocol_tungstenite_client.rs

```diff
@@ -332,11 +548,119 @@ impl SmartTransferProtocolClient for WsSmartTransferProtocolClient {
+        listener.begin();
+
+        let src_head = src
+            .as_metadata_chain()
+            .get_ref(&BlockRef::Head)
+            .await
+            .int_err()?;
+
+        let size_estimation =
+            match prepare_dataset_transfer_estimaton(src.as_metadata_chain(), &src_head, dst_head)
+                .await
+            {
+                Ok(size_estimation) => size_estimation,
+                Err(e) => return Err(SyncError::Internal(e.int_err())),
+            };
+
+        let num_blocks = size_estimation.num_blocks;
+        if num_blocks == 0 {
+            return Ok(SyncResult::UpToDate);
+        }
+
+        let mut push_url = http_dst_url.join("push").unwrap();
+        let push_url_res = push_url.set_scheme("ws");
+        assert!(push_url_res.is_ok());
+
+        tracing::debug!("Connecting to push URL: {}", push_url);
+
+        let mut ws_stream = match connect_async(push_url).await {
+            Ok((ws_stream, _)) => ws_stream,
+            Err(e) => {
+                tracing::debug!("Failed to connect to push URL: {}", e);
+                return Err(SyncError::Internal(e.int_err()));
+            }
+        };
+
+        match self
+            .push_send_request(&mut ws_stream, size_estimation, dst_head)
+            .await
+        {
+            Ok(_) => {}
+            Err(e) => {
+                tracing::debug!("Push process aborted with error: {}", e);
+                return Err(SyncError::Internal(e.int_err()));
+            }
+        };
+
+        match self
+            .push_send_metadata_request(&mut ws_stream, src, &src_head, dst_head)
+            .await
+        {
+            Ok(_) => {}
+            Err(e) => {
+                tracing::debug!("Push process aborted with error: {}", e);
+                return Err(SyncError::Internal(e.int_err()));
+            }
+        };
+
+        let missing_objects =
+            match collect_missing_object_references_from_interval(src, &src_head, dst_head).await {
+                Ok(object_references) => object_references,
+                Err(e) => {
+                    tracing::debug!("Push process aborted with error: {}", e);
+                    return Err(SyncError::Internal(e.int_err()));
+                }
+            };
+
+        let push_objects_response = match self
+            .push_send_objects_request(&mut ws_stream, missing_objects)
+            .await
+        {
+            Ok(push_objects_response) => push_objects_response,
+            Err(e) => {
+                tracing::debug!("Push process aborted with error: {}", e);
+                return Err(SyncError::Internal(e.int_err()));
+            }
+        };
+
+        use futures::stream::{StreamExt, TryStreamExt};
+        futures::stream::iter(push_objects_response.object_transfer_strategies)
+            .map(Ok)
+            .try_for_each_concurrent(
+                /* limit */ 4, // TODO: external configuration?
```

Comment:
Yes, let's make it a struct field to be configured when creating the client.

Reply:
Externalized `ObjectTransferOptions` struct. By default, it is taking the value of [available parallelism](https://doc.rust-lang.org/std/thread/fn.available_parallelism.html) on the target platform.